File: tech_question.py
"""
Step 1: First Principles
- Create a Pursuit Algorithm that will will chase a target in a 2D plane.
- The target will move in a straight line at a constant speed.

Step 2: Make it better
- How can we adjust velocity
- How can we adjust the control of the pursuer with constraints

Step 3: Make it even better
- How can we implement this in a class

"""
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import copy

from typing import List

CAPTURE_DISTANCE: float = 2.0
EVADER_VELOCITY: float = 5.0

"""
Implement the algorithm here
"""


def pure_pursuit_algorithm(pursuer_position: np.ndarray,
                           target_position: np.ndarray) -> tuple:
    """
    Implement the pursuit algorithm
    It will return the heading command for the pursuer
    and a velocity command (magnitude) for now keep velocity constant

    """
    # Calculate the heading command, keep velocity constant
    heading_cmd = 0.0
    vel_cmd = 10.0

    return vel_cmd, heading_cmd


def is_close(pursuer_position: np.ndarray,
             target_position: np.ndarray) -> bool:
    """
    Check if the pursuer is close to the target
    return True if close
    """

    distance = 100
    return False


def move(position: np.ndarray,
         velocity: np.ndarray,
         heading_cmd_rad: float,
         dt: float) -> np.ndarray:
    """
    Move the position of the target based on the velocity and time.
    Position is a 2d array that consists of x, y, and heading
    """
    position[0] += velocity * np.cos(position[2]) * dt
    position[1] += velocity * np.sin(position[2]) * dt
    position[2] += heading_cmd_rad * dt

    return position


def plot_simulation(history_pursuer: List[np.ndarray],
                    history_evader: List[np.ndarray]) -> None:
    """
    Plot the simulation
    """
    fig, ax = plt.subplots()
    ax.plot([x[0] for x in history_pursuer],
            [x[1] for x in history_pursuer],
            label='Pursuer',
            color='red')
    ax.plot([x[0] for x in history_evader],
            [x[1] for x in history_evader],
            label='Evader',
            color='blue')

    ax.legend()

    # plots the velocity and heading
    fig, ax = plt.subplots()
    ax.plot([np.rad2deg(x[2]) for x in history_pursuer],
            label='Pursuer Heading',
            color='red')
    ax.plot([np.rad2deg(x[2]) for x in history_evader],
            label='Evader Heading',
            color='blue')
    ax.legend()


def sim() -> None:

    dt: float = 0.05
    start_time: float = 0.0
    end_time: float = 10.0

    pursuer_position: np.array = np.array([-15.0,
                                           -15.0,
                                           np.deg2rad(45.0)])

    evader_position: np.array = np.array([0.0,
                                          0.0,
                                          np.deg2rad(0.0)])

    num_steps = int((end_time - start_time) / dt)

    history_pursuer = []
    history_evader = []

    for i in range(num_steps):

        # Implement the correct pursuit algorithm HERE
        vel_cmd, heading_cmd_rad = pure_pursuit_algorithm(
            pursuer_position, evader_position)

        pursuer_position = move(pursuer_position,
                                vel_cmd,
                                heading_cmd_rad,
                                dt)

        evader_position = move(evader_position,
                               EVADER_VELOCITY,
                               0.0,
                               dt)

        history_evader.append(copy.deepcopy(evader_position))
        history_pursuer.append(copy.deepcopy(pursuer_position))

        # Implement the correct logic to check if the pursuer has captured the evader
        if is_close(pursuer_position, evader_position):
            print("Captured")
            break

    # logging data to plot
    df = pd.DataFrame(history_pursuer, columns=['x', 'y', 'heading'])
    df.to_csv('pursuer.csv', index=False)

    df = pd.DataFrame(history_evader, columns=['x', 'y', 'heading'])
    df.to_csv('evader.csv', index=False)

    plot_simulation(history_pursuer, history_evader)
    plt.show()


if __name__ == '__main__':
    sim()

File: tech_question.cpp
#include <stdio.h>
#include <iostream>
#include <vector>
#include <cmath>
#include <math.h>

float CAPTURE_DISTANCE = 2.0;
float EVADER_VELOCITY = 5.0;

//// -------------- IMPLEMENTATION -------------------------------- ////
// Implement the pure pursuit algorithm
std::vector<float> pure_pursuit_algorithm(
    std::vector<float> evader_position, 
    std::vector<float> pursuer_position, float dt)
{
    float vel_cmd = 10.0;
    float heading_cmd = 0.0;
    // Calculate the heading command keep velocity constant
    return {vel_cmd, heading_cmd};
}

// Compute the distance between two points and returns true if within CAPTURE_DISTANCE
bool is_close(std::vector<float> evader_position, 
    std::vector<float> pursuer_position, float capture_distance)
{
    float distance = 100.0;
    return false;
}


/// --------------Helper functions ----------------------------------- ///    
float rad2deg(float rad)
{
    return rad * 180.0 / M_PI;
}

float deg2rad(float deg)
{
    return deg * M_PI / 180.0;
}

std::vector<float> move(std::vector<float> position, 
    float velocity, float heading_cmd, float dt)
{

    position[0] += velocity * cos(position[2]) * dt;
    position[1] += velocity * sin(position[2]) * dt;
    position[2] += heading_cmd; //* dt;

    return position;
}

void save_to_csv(std::vector<std::vector<float>> evader_positions, 
    std::vector<std::vector<float>> pursuer_positions)
{
    FILE *f = fopen("pursuit.csv", "w");
    fprintf(f, "evader_x,evader_y,evader_theta,pursuer_x,pursuer_y,pursuer_theta\n");
    for (int i = 0; i < evader_positions.size(); i++)
    {
        fprintf(f, "%f,%f,%f,%f,%f,%f\n", 
            evader_positions[i][0], evader_positions[i][1], evader_positions[i][2],
            pursuer_positions[i][0], pursuer_positions[i][1], pursuer_positions[i][2]);
    }

    fclose(f);
}

// Main function
int main()
{
    float start_time = 0.0;
    float end_time = 10.0;
    float dt = 0.05;
    int num_steps = int((end_time - start_time) / dt);

    std::vector<float> pursuer_position = {
        -15.0, // x
        -15.0, // y
        deg2rad(45.0) // theta
    };

    std::vector<float> evader_position = {
        0.0, // x    
        0.0, // y
        deg2rad(0.0) // theta
    };


    // Used to store the positions of the evader and pursuer
    std::vector<std::vector<float>> evader_positions;
    std::vector<std::vector<float>> pursuer_positions;

    // Simulator loop
    for (int i = 0; i <= num_steps; i++)
    {
        // IMPLEMENT PURSUIT ALGORITHM
        std::vector<float> control_cmd = pure_pursuit_algorithm(
            evader_position, pursuer_position, dt);

        pursuer_position = move(pursuer_position, 
            control_cmd[0], control_cmd[1], dt);

        evader_position = move(evader_position, 
            EVADER_VELOCITY, 0.0, dt);

        evader_positions.push_back(evader_position);
        pursuer_positions.push_back(pursuer_position);

        // IMPLEMENT DISTANCE CHECK
        if (is_close(evader_position, pursuer_position, CAPTURE_DISTANCE))
        {
            printf("Captured at step %d\n", i);
            break;
        }
        
        
    }

    save_to_csv(evader_positions, pursuer_positions);

}
